kid0012.md

```diff
@@ -0,0 +1,60 @@
+8. If there is duplicity, super watchers detect and report it within 2 minutes of its manifestation in diverged witness state.
+
+## Requirements
+
+9. Witnesses MUST implement an endpoint that describes which of the following features they support, so that their behavior is discoverable. (Or should the TEL of the witness simply contain a commitment to support certain features?)
+
+10. Witnesses SHOULD support an efficient, comprehensive polling feature required by super watchers. Essentially, super watchers will ask, "What changes have you recorded since your &lt;checkpoint X&gt;?", and witnesses implementing the feature MUST respond with a batch of changes.
+
+    ![diagram of polled witnesses](assets/polled-witnesses.png)
+
+    This is probably implemented by having witnesses rotate to a new "current checkpoint ID" once every minute. All changes that the witness records in the next minute are associated with that new checkpoint ID. The witness maintains a circular index of the associations between changes and checkpoints for at least 2 hours (120 checkpoints).
+
+    Witnesses SHOULD also support a variation on this polling request that constrains the AIDs of interest: "What changes have you recorded since your &lt;checkpoint X&gt;, for the following AIDs: {AID1, AID2, AID3...}?"
+
+11. Witnesses SHOULD throttle polling requests.
+
+12. Witnesses MAY also restrict who can poll. Exactly how they do this is out of scope. One obvious way to restrict is by requiring polling requests to be signed by an AID on an allow list.
+
+13. Witnesses MAY support a push notification feature as well. If they do, they allow themselves to be configured so that they push notifications of changes in the KEL/TEL of any of their serviced AIDs to an arbirary endpoint (a super watcher).
+
+![pushing witnesses](assets/pushing-witnesses.png)
+
+The super watcher benefits because it learns of changes with lower latency and greater efficiency than it would if it polled. However, the witness accepts a burden to provide this feature. For this reason, witnesses will probably only allow the feature if the watcher in question belongs to the same or an allied owner.
+
+14. If witnesses support this push feature, they SHOULD push to only one location, delivery SHOULD be best-effort, and witnesses SHOULD contain no retry logic or state tracking. This preserves the simplicity of witnesses, and their focus on recording data, not moving it.
+
+15. If witnesses support this push feature, they SHOULD also support aggregate push. This is a network optimization, where updates that occur within a configurable timeframe (e.g., 10 seconds) are all pushed as a single notification.
+
+16. Duplicity detection. A watcher that detects duplicity MUST immediately notify all witnesses for the associated AID, and witnesses MUST expose an endpoint to receive such reports. (What else should be done to react to duplicity?)
```

Comment:
The controller should have witnesses and watchers. So the separation of responsibility seems to me:
The watcher sends a notification of duplicity to anyone who is using that watcher (including the controller).
Then it is the responsibility of the controller to take action (rotate keys, rotate witnesses, etc.) to reestablish a non-duplicitous state.

Reply:
I agree that what you have said is desirable, and I considered saying something similar in my draft. What stopped me was this question: "*HOW*, exactly, does a witness or a watcher contact a controller?" The problem is that A) controllers of AIDs may not be directly reachable (although we hope they use KERIA-style agents [~=DIDComm mediators, ~=Trust Spanning Protocol construct with a name I'm forgetting] so messages can be sent even when the controllers are offline); and B) given an endpoint where controllers can be contacted, what protocol do we speak to tell them about the problem? I'm not asking about the low-level protocol (DIDComm, TSP, etc); I'm talking about the protocol that says, "I'm reporting a duplicity event in the pre-agreed format for such things. Here you go." No such protocol exists, AFAIK.

So: if we make such a statement here, I feel like we incur an implicit duty to say how it should be done. Want to write another KID? :-)

Reply:
That's a great point. I believe Sam implemented several ['Likely Duplicitous' features](https://github.com/WebOfTrust/keripy/blob/6ab521ec83f1bfd67882199fea4c5f45d91304c4/src/keri/kering.py#L398) in KERIpy that I need learn about. And then the community should have a discussion about previously discussed protocols. Certainly the new IPEX protocol/messaging that Phil, Rodo, and others are working on should help inform the discussion as well.